[Zhu Yufan] iP

Yufannnn opened the pull request:
# Mr. Bear
> “Your mind is for having ideas, not holding them.”  – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Mr. Bear is a **convenient and flexible** software that helps you to manage and track all your tasks. It's,
+ text-based
+ easy to learn
+ ~~FAST~~ *SUPER FAST* to use

**And it is ***FREE!!!***

All you need to do is,
1. download it from [here](https://github.com/Yufannnn/ip/releases). 
2. double-click it. 
3. add your tasks. 
4. let it manage your tasks for you 😉

If you are not sure how to use it, check out the [User Guide](https://yufannnn.github.io/ip/)! 

Features:

- [X] Graphic User Interface
- [X] Add four type of tasks (todo, deadline, event, fixed duration)
- [X] Search, Sort, Find tasks
- [X] Mark, Unmark, Delete, massDelete tasks
- [ ] Undo Command (Partially functional and comming soon...)

Yufannnn committed:
added fruits text file

Yufannnn committed:
removed fruits.txt

Yufannnn committed:
Add Greet, Echo, and Exit Features

Yufannnn committed:
Added Add, List feature

Yufannnn committed:
Added Mark, Unmark feature

Yufannnn committed:
Added Todo, Event, Deadline Tasks

Yufannnn committed:
Added Error Handling Mechanism

Yufannnn committed:
Added Delete feature

Yufannnn committed:
Add TextUiTesting feature

Yufannnn committed:
Simplified DukeException

Yufannnn committed:
Added Save Feature

Yufannnn committed:
Add Dates and Times features

Yufannnn committed:
Merge branch 'branch-Level-7'

Yufannnn committed:
Merge branch 'branch-Level-8'

Yufannnn committed:
Resovle the Compatibility Issue Between Level-7 and Level-8

Yufannnn committed:
Modified the Code to be more OOP, reaching the minmal requirement

Yufannnn committed:
Modified the structure of the code, achieved the stretch level of OOP, changed names to fit the suggested standard and removed obsolete Commands.

Yufannnn committed:
Added Gradle Wrapper to the Project, achieving the second level of being about to execute unit test.

Yufannnn committed:
Added the Jar Featurea and Fixed some issue with build.gradle

Yufannnn committed:
Add JavaDoc comments to the code, acheving the stretch level.

Yufannnn committed:
Added Gradle Checkstyle feature and config file. Fixed the coding standard issues.

Yufannnn committed:
Added Find feature.

Yufannnn committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	data/tasks.txt

Yufannnn committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/display/Ui.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/EventTask.java
#	src/test/java/duke/parser/ParserTest.java

Yufannnn committed:
Resolved conflicts when merging the branches. Fixed Styling issue and added more JavaDoc

Yufannnn committed:
Added Search Feature, allowing user to search deadline and event happened on a given date. (Achieving Streach Goal of Level 8)

Yufannnn committed:
Updated ReadMe.md

Yufannnn committed:
Added GUI to the App, changed the name of the project from Duke to Mr. Bear. Updated readMe.md

Yufannnn committed:
Allowed to find multiple keyword at the same time using Vararg

Yufannnn committed:
Release the v2.0 version of the ChatBot.

Yufannnn committed:
Improve Code Quality by break down long methods. Fixed the bug of not reading eventTask correctly.

Yufannnn committed:
Added Assertion Feature

The previous version of Mr. Bear Does not have any checking mechanisms
to verify the format of the event stored in the storage file as I
assumed that events would be written to the file in the correct format.
However, this cannot be guaranteed and should any event be stored in
the wrong format and loaded incorrectly, the whole program would
potentially crash. Therefore, it is the necessary to add an
assertion feature to verify the assumption is indeed true.

Add assertion to check whether the event type are of the format [T],
[D], [E], representing TodoTask, DeadlineTask, and EventTask
respectively. Add assertion to check whether the isDone status is of
the format [ ], or [X]. If not, terminate the program and display the
relevant error message.

Yufannnn committed:
Merge pull request #2 from Yufannnn/branch-A-Assertions

Added Assertion Feature

Yufannnn committed:
Merge branch 'master' into branch-A-CodeQuality

Yufannnn committed:
Merge pull request #3 from Yufannnn/branch-A-CodeQuality

Improve Code Quality

Yufannnn committed:
Added Help Feature

The current version of Mr. Bear does not have an in-app help menu that allows user to check the format of the their command. Although it does have error prompt to notify the user when the input invalid command, it is still rather inconvenient.

Therefore, I added an help command to allow the user to check all the available command and their respective functionalities by displaying it in a dialog box.

Yufannnn committed:
Added Update Task Description Feature

In order to update the description of a task in the current version of Mr. Bear, the user has no choice but to delete the task first and added a new one with the intended description, which is very tedious.

Therefore, a direct update function is added to allow the users to alter the description easily and the help menu is also changed.

Yufannnn committed:
Added Reminder, Mass Delete Feature

Added Reminder feature with the help of Java Stream to display all the deadline events in sorted by their respective deadline.
Added MassDelete feature that allows user to delete all the tasks that have been done in one go.
Modified the constant names to follow the coding standard.

Yufannnn committed:
Update README.md

Yufannnn committed:
Improved Code Quality

Improve the overall code quality, make the code more readable. Added more javaDoc when missing.

Yufannnn committed:
Merge branch 'master' of https://github.com/Yufannnn/ip

Yufannnn committed:
Improve Code Quality for all Command Classes

Yufannnn committed:
Modified the Original Search Feature to View Schedule Feature

Yufannnn committed:
Add FindFreeTime Feature

This allows user to find a free date in the next month.

Yufannnn committed:
Update README.md

Yufannnn committed:
Update README.md

Yufannnn committed:
Improve Code Quality for Multiple Classes

Yufannnn committed:
Add Inline Comment to Classes

Add inline comments to all the method and fields to make it easier to understand.

Yufannnn committed:
Allows the Task time to be more Specific

The current version of Mr. Bear only supports time as specific as a single date. I updated to allow it to be able to be specific to the minute and also updated the help information.

Yufannnn committed:
Added Fixed Duration Task

Added a new type of Task, which is the Fixed Duration task, improve the GUI slightly.

Yufannnn committed:
Add Sort Feature

Yufannnn committed:
Set up CI

Use GitHub Actions to set up Continuous Integration (CI).

The workflow specified by this .yml file is a good candidate for this project. The last three segments are related to I/O redirection tests; can be deleted if not applicable to your project.

Yufannnn committed:
Merge pull request #4 from Yufannnn/branch-A-CI

Set up CI

Yufannnn committed:
Revert "Set up CI"

Yufannnn committed:
Merge pull request #5 from Yufannnn/revert-4-branch-A-CI

Revert "Set up CI"

Yufannnn committed:
Set A-CI

Use GitHub Actions to set up Continuous Integration (CI).

The workflow specified by this .yml file is a good candidate for this project. The last three segments are related to I/O redirection tests; can be deleted if not applicable to your project.

Yufannnn committed:
Set up CI

Use GitHub Actions to set up Continuous Integration (CI).

The workflow specified by this .yml file is a good candidate for this project. The last three segments are related to I/O redirection tests; can be deleted if not applicable to your project.

Yufannnn committed:
Create _config.yml

Yufannnn committed:
Update _config.yml

Yufannnn committed:
Add more Test Cases

Yufannnn committed:
Merge branch 'Yufannnn-A-CI'

Yufannnn committed:
Update README.md

Yufannnn committed:
Add User Guide

Yufannnn committed:
Update README.md

Yufannnn committed:
Update README.md

Yufannnn committed:
Update README.md

kyueran commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,166 @@
+     * @throws DukeException Throws exception when invalid input is given
+     */
+    public static Command parse(String input) throws DukeException {
+        Matcher instructionExtractor = extractInstructionAndInformation(input);
+        String instructionTag = instructionExtractor.group("instructionTag").strip();
+        String information = instructionExtractor.group("information").strip();
```
Good job having detailed names for the variables in your parse function!

kyueran commented on src/main/java/duke/storage/Decoder.java:
```diff
@@ -0,0 +1,109 @@
+        if (isDone) {
+            deadline.markAsDone();
+        }
+
```
Should we leave line breaks in functions?

kyueran reviewed (commented):
LGTM, very well done PR with a lot of great additions to Duke!

jugsliao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,74 @@
+ * The main Duke class to run
+ */
+public class Duke {
+    private final Storage storage;
```
Perhaps the storage variable name could be in caps because it is final. Similarly for other final variables.

jugsliao commented on src/main/java/duke/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,80 @@
+        if (!(obj instanceof AddTaskCommand)) {
+            return false;
+        }
+        AddTaskCommand ddlObj = (AddTaskCommand) obj;
```
perhaps you could spell "ddlObj" out so it is clearer to the reader of your code.

jugsliao commented on src/main/java/duke/command/HelpCommand.java:
```diff
@@ -0,0 +1,175 @@
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+public class HelpCommand extends Command {
```
I like how you you implemented this class to make your app more interesting.

jugsliao commented on src/main/java/duke/command/HelpCommand.java:
```diff
@@ -0,0 +1,175 @@
+                "place on the given day, type \"help date\" to check all the available date format");
+
+    private final String command;
+        private final String description;
```
Perhaps description should be in caps.

jugsliao commented on src/main/java/duke/parser/Decipherer.java:
```diff
@@ -0,0 +1,306 @@
+ * /
+ */
+public class Decipherer {
+    private static final Pattern emptyStringChecker = Pattern.compile("\\S.*+");
```
Perhaps variable name should be caps instead of camelcase because its final.

jugsliao commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,90 @@
+     * @param indentSpace The length of the indent space to be set
+     */
+    public Ui(int horizontalLineLength, int indentSpace) {
+        this.horizontalLineLength = horizontalLineLength;
```
I like how your variable name is comprehensive and clear.

jugsliao reviewed (commented):
Overall, I found your code very comprehensive yet not so complex that it is difficult to read. Good job on writing such good quality code. There are a few standard coding violations but the naming of all your methods, class and variable names make sense.

Yufannnn committed:
Remove Undo Structure

The current version is to bug-prune so I decided to remove it.

Yufannnn committed:
Merge branch 'master' of https://github.com/Yufannnn/ip

# Conflicts:
#	docs/README.md

Yufannnn committed:
Update README.md

Yufannnn committed:
Add Ui.png

Yufannnn committed:
Delete UI.png

Yufannnn committed:
Update README.md

Bentimate commented on src/main/java/duke/command/FindFreeTimeCommand.java:
```diff
@@ -0,0 +1,111 @@
+            else if (task instanceof EventTask) {
+                if (isEventTaskScheduledOnDate(date, (EventTask) task)) {
+                    return false;
+                }
```
Perhaps the conditions for the if statements can be combined to avoid too many indentations

Bentimate commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,223 @@
+     */
+    private void createParentFolderIfNotExists() {
+        // check if the folder exists
+        if (!this.folder.toFile().exists()) {
+            // if it does not, create the folder using mkdirs()
+            this.folder.toFile().mkdirs();
+        }
```
Perhaps this can be structured as a guard clause instead?

Bentimate reviewed (commented):
Very easy to read and logical in structure, and overall good code quality!

Yufannnn committed:
Improve Code Quality

An additional round of code Quality improvement including fix styling errors, break down long methods, and add more javaDoc and comments where missing.

Yufannnn committed:
Update gradle.yml

Yufannnn committed:
Improve Code Quality

Yufannnn committed:
Merge branch 'master' of https://github.com/Yufannnn/ip

Yufannnn committed:
Update README.md

Yufannnn committed:
Updated README.md

Yufannnn committed:
Code Styling Refinement

Yufannnn committed:
Updated build.gradle

Yufannnn committed:
Update gradle.yml

Yufannnn committed:
Update README.md
